internal/cephfs/nodeserver.go

```diff
@@ -127,10 +129,54 @@ func maybeUnlockFileEncryption(
 	stagingTargetPath string,
 	volID fsutil.VolumeID,
 ) error {
-	if volOptions.IsEncrypted() {
-		log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
+	if !volOptions.IsEncrypted() {
+		return nil
+	}
+
+	// Define Mutex Lock variables
+	lockName := string(volID) + "-mutexLock"
+	lockCookie := lockName + "-coockie"
+	lockDesc := "Lock for " + string(volID)
+	lockDuration := 150 * time.Second
+	var flags byte = 0
 
-		return fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
+	log.DebugLog(ctx, "Creating lock for the following volume ID %s", volID)
+
+	ioctx, err := volOptions.GetConnection().GetIoctx(volOptions.MetadataPool)
+	if err != nil {
+		log.ErrorLog(ctx, "Failed to create ioctx: %s", err)
+	}
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, lockCookie, lockDesc, lockDuration, &flags)
+	if res != 0 {
```

Comment:
https://github.com/ceph/go-ceph/blob/1046b034a1f618f67acd3c6523482917e27c7113/rados/ioctx.go#L545-L558 Looking at this one, we need to return an error in case of EBUSY and EEXIST at line 161 it return nil, we need to have return in those cases as well, @nixpanic can you please check this one?

Reply:
EBUSY and EEXIST are expected return values, not an error while calling the function. I do not think it should return an error in those cases.

Reply:
sorry i meant not in the go-ceph from here cephcsi should return an error so that if the lock is held on one node we should keep retrying on another node until it's done, or we just return nil (which is happening here) assuming that the other client will unlock fscrypt?

Reply:
ah, yes, now I see what you mean. If EBUSY/EEXISTS is returned from `LockExclusive()`, `err` will be `nil`. That `err` is then returned on line 162, and the caller of this function assumes all is OK (because `err` is `nil`). This is not correct, and instead of returning `err` on line 162, a new error should be returned. Ideally each `case` in the below switch sets `err` to a new unique value (mentioning something similar as what is in the debug log).